File: 22-generate-parentheses/22-generate-parentheses.java
class Solution {
    private List<String> res;
    private int n;
    public List<String> generateParenthesis(int n) {
        res = new ArrayList<>();
        this.n = n;
        backtrack(new StringBuilder(), 0, 0);
        return res;
    }
    
    private void backtrack(StringBuilder comb, int open, int close) {
       if (open == n && close == n) {
            res.add(comb.toString());
       }
       if (open < n) {
           comb.append('(');
           backtrack(comb, open+1, close);
           comb.deleteCharAt(comb.length()-1);
       }
       if (close < open) {
           comb.append(')');
           backtrack(comb, open, close+1);
           comb.deleteCharAt(comb.length()-1);
       }
    }
}

File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        backtrack(n, res, 0, 0, "");
        return res;
    }

private:
    void backtrack(int n, vector<string>& res, int open, int close, string comb) {
        if (comb.size() == 2*n) {
            res.push_back(comb);
            return;
        }
        if (open < n) {
            backtrack(n, res, open+1, close, comb+'(');
        }
        if (close < open) {
            backtrack(n, res, open, close+1, comb+')');
        }
    }
    
};